binding/c: autogen rma api

hzhou opened the pull request:
## Pull Request Description

Let python scripts to generate RMA API functions. MPI files in `src/mpi/rma` were removed and `mpir` routines (if any) are moved into `rma_impl.c`. New binding files are generated in `src/binding/c/rma/`.

Removed mpi errnames (e.g. **mpi_get) from src/mpi/errhan/errnames.txt because now script generates them. The script's versions are slightly different from original mostly because it uses %i for ranks instead of %d, and %L for MPI_Aint instead of %d.          

Scripts generates validation based on parmeter types. Potentially it may generate more validations than previously does and in some cases inapproprate ones especially when certain parameters are allowed to be
NULL. We use custom "error_check" and `.skip validate-KIND` to fix those in `src/binding/c/rma_api.txt`.

Added validation checks for enums of `assert` and `locketype`. `CHECKENUM` checks the param against a list of enum values. `CHECKMASK` checks the param against a list mask positions (that can be `or`ed together).

Man pages are automatically generated now. Slight differences or errors may slip in.
[skip warnings]

## Expected Impact

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] You or your company has a signed contributor's agreement on file with Argonne
* [x] For non-Argonne authors, request an explicit comment from your companies PR approval manager

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

minsii reviewed (commented):
I need spend more time to better compare the autogened RMA files v.s. the original files.

One quick question is, why do we change the error messages in `src/mpi/errhan/errnames.txt`?

hzhou commented:
> I need spend more time to better compare the autogened RMA files v.s. the original files.
> 
> One quick question is, why do we change the error messages in `src/mpi/errhan/errnames.txt`?

The script generates error messages, e.g. `**mpi_put %W...` based on parameter types. It end up often with slightly different format than previously hand maintained ones. Rather than manually keep them in sync, I let the script to generate `src/binding/c/errnames.txt` so it will always be consistent. As a result, we no longer need manually maintained entries in `src/mpi/errhan/errnames.txt`, thus deleting them in the PR.

For RMA, I also added standard way of checking enum values, e.g. `assert` and `locktype` -- refer to `CHECKENUM in `src/binding/c/rma_api.txt`, which also generates errnames in certain way, that's why those two errnames are modified in this PR.

minsii commented:
I see. Can you please also add a summary of key changes compared to the original files in commit message? I.e., what are the new features added in the script and used for RMA files.

hzhou commented:
> I see. Can you please also add a summary of key changes compared to the original files in commit message? I.e., what are the new features added in the script and used for RMA files.

Sure. I'll make sure to do that later.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+MPI_Win_allocate_shared:
+    .desc: Create an MPI Window object for one-sided communication and shared memory access, and allocate memory at each process
+    .seealso: MPI_Win_allocate MPI_Win_create MPI_Win_create_dynamic MPI_Win_free MPI_Win_shared_query
+    .impl: mpid
```
In `win_allocate_shared.c`, error `.N MPI_ERR_INFO` and `.N MPI_ERR_SIZE` are lost. For invalid size argument, I think it is better to return `MPI_ERR_SIZE` but not `MPI_ERR_ARG`. Same for `disp`.

minsii replied:
The code format in `win_allocate_shared.c` seems weird (see `MPIR_OBJ_PUBLISH_HANDLE ` line)
```
    if (win_ptr) {
    /* Initialize a few fields that have specific defaults */
    win_ptr->name[0] = 0;
    win_ptr->errhandler = 0;
        MPIR_OBJ_PUBLISH_HANDLE(*win, win_ptr->handle);
    }
```

minsii replied:
I see in many files, the header comment is lost. For example, in `win_allocate_shared.c` there was a comment:
```
This is a collective call executed by all processes in the group of comm. On
each process i, it allocates memory of at least size bytes that is shared among
all processes in comm, and returns...
```

hzhou replied:
> In `win_allocate_shared.c`, error `.N MPI_ERR_INFO` and `.N MPI_ERR_SIZE` are lost. For invalid size argument, I think it is better to return `MPI_ERR_SIZE` but not `MPI_ERR_ARG`. Same for `disp`.
> 

Let me add `MPIR_ERRTEST_RMASIZE` and `MPIR_ERRTEST_DISPUNIT` for them, will update.

hzhou replied:
> 
> 
> The code format in `win_allocate_shared.c` seems weird (see `MPIR_OBJ_PUBLISH_HANDLE ` line)
> 
> ```
>     if (win_ptr) {
>     /* Initialize a few fields that have specific defaults */
>     win_ptr->name[0] = 0;
>     win_ptr->errhandler = 0;
>         MPIR_OBJ_PUBLISH_HANDLE(*win, win_ptr->handle);
>     }
> ```

Added commit to fix python script

hzhou replied:
> 
> 
> I see in many files, the header comment is lost. For example, in `win_allocate_shared.c` there was a comment:
> 
> ```
> This is a collective call executed by all processes in the group of comm. On
> each process i, it allocates memory of at least size bytes that is shared among
> all processes in comm, and returns...
> ```

Oh I see. That notes was before the parameter description block, which my original scraping script didn't anticipate. They really should be after the parameter block as all the other man pages. I'll run a script to check all the missing ones. To be updated.

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    .seealso: MPI_Win_create_dynamic MPI_Win_detach
+    .impl: mpid
+{ -- early_return --
+    if (size == 0)
```
In `win_attach`, `MPI_ERR_SIZE` error check is lost

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    .desc: Create an MPI Window object for one-sided communication
+    .seealso: MPI_Win_attach MPI_Win_detach MPI_Win_allocate MPI_Win_allocate_shared MPI_Win_create MPI_Win_free
+    .impl: mpid
+/*
```
In the autogen file of `win_create_dynamic.c`, why the input parameter is `- comm - intra-communicator (handle)`? We don't specify intra or inter in any other RMA files.

minsii replied:
I think `MPIR_Err_return_win` should be `MPIR_Err_return_comm`  at `fn_fail` in `win_create_dynamic.c`, because the win_ptr is not created yet.
Same issue in `win_create.c`.

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    .impl: mpid
+{ -- error_check --
+    CHECKMASK: assert, assert, MPI_MODE_NOSTORE MPI_MODE_NOPUT MPI_MODE_NOPRECEDE MPI_MODE_NOSUCCEED
+}
```
```
   Notes:
   The 'assert' argument is used to indicate special conditions for the
   fence that an implementation may use to optimize the 'MPI_Win_fence'
   operation.  The value zero is always correct....
```
The above note in `fence.c` is lost.

minsii replied:
`MPI_ERR_ARG` is used but not in header comment of `fence.c`

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+MPI_Win_free:
+    .desc: Free an MPI RMA window
+{
+    if (MPIR_Process.attr_free && win_ptr->attributes) {
```
Note below is lost in `win_free.c`
```
   Notes:
   If successfully freed, 'win' is set to 'MPI_WIN_NULL'.
```

minsii replied:
Note is lost in `win_get_group.c` too.

minsii replied:
In `win_get_group.c`, do we want to replace ` *group = group_ptr->handle;` with `MPIR_OBJ_PUBLISH_HANDLE`?

hzhou replied:
> 
> 
> Note is lost in `win_get_group.c` too.

Added

hzhou replied:
> 
> 
> In `win_get_group.c`, do we want to replace ` *group = group_ptr->handle;` with `MPIR_OBJ_PUBLISH_HANDLE`?

Why not?

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    .impl: mpid
+    .earlyreturn: pt2pt_proc_null
+{ -- error_check --
+    CHECKMASK: assert, assert, MPI_MODE_NOCHECK
```
Note is lost in `win_lock.c`, `win_post.c`, `win_start.c`.

hzhou replied:
Added.

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    .desc: Query the size and base pointer for a patch of a shared memory window
+    .seealso: MPI_Win_allocate_shared
+    .impl: mpid
+
```
In `win_shared_query`, The original description for output `baseptr` is useful. Can we keep it?
```
- baseptr - base pointer in the calling process'' address space of the shared
  segment belonging to the target rank.
```

hzhou replied:
Parameter description is from  `src/binding/c/mpi_standard_api.txt`, which is kept in sync with MPI standard, and I think we should keep it this way.

We can add a note for any specifics that is missing from standard description or that we want to emphasize. Would you like me to do so?

minsii replied:
maybe we can move the description to the note paragraph? No need to hack the auto-sync file.

hzhou replied:
> maybe we can move the description to the note paragraph? No need to hack the auto-sync file.

👍

hzhou replied:
fixed

hzhou reviewed (commented):
A batch -- will continue

hzhou reviewed (commented):
Addressed most of the comments. Will update soon.

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
@minsii Ready for review.

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    types, or Byte as specified in Section 5.9.2 on page 176. The origin and result
+    buffers (origin_addr and result_addr) must be disjoint.
+*/
+{ -- error_check --
```
> Do we have a legitimate reason to check datatype when `count` is `0`? For example, is there standard text that implies e.g `MPI_Put(NULL, 0, MPI_DATATYPE_NULL, target, 0, 0, MPI_DATATYPE_NULL, win)` is wrong? On the other hand, adding special rules for RMA "userbuffer" check adds complexity to the python script that we should avoid unless necessary.

I don't think the above usage is a valid case in RMA. Unlike pt2pt, zero-count rma cannot be used as synchronization.  The only special case is fetch-like atomics with `NO_OP` where the origin_datatype can be `MPI_DATATYPE_NULL`. Thus, I don't think we need add the unnecessary  `if (origin_count > 0) {...}` code .

minsii replied:
> We have to manually annotate on which function need `PT2PT|COLL|RMA` specific logging and the rules from function names seems overly complex to add into the python scripts. Since I don't think there is anything specific about these logging -- probably from now removed `rlog` -- I took the simplified way.
> 
> Do you recall what was specific about these per-(pt2pt|coll|rma) logging?
> 
> The separate commit will essentially just rename the macros in those mpi functions that immediately get deleted in the next commit. If you insist, I'll do it.

In rlog, all are redirected to the same func. Maybe we can safely remove them. But should the change be in a sperate commit? 
```
#define RLOG_PT2PT_FUNC_ENTER(a)     RLOG_FUNC_ENTER(a)
#define RLOG_PT2PT_FUNC_EXIT(a)      RLOG_FUNC_EXIT(a)
#define RLOG_COLL_FUNC_ENTER(a)      RLOG_FUNC_ENTER(a)
#define RLOG_COLL_FUNC_EXIT(a)       RLOG_FUNC_EXIT(a)
#define RLOG_RMA_FUNC_ENTER(a)       RLOG_FUNC_ENTER(a)
#define RLOG_RMA_FUNC_EXIT(a)        RLOG_FUNC_EXIT(a)
```

hzhou replied:
Ok, I will add the special condition for RMA in the python script.

hzhou replied:
> > We have to manually annotate on which function need `PT2PT|COLL|RMA` specific logging and the rules from function names seems overly complex to add into the python scripts. Since I don't think there is anything specific about these logging -- probably from now removed `rlog` -- I took the simplified way.
> > Do you recall what was specific about these per-(pt2pt|coll|rma) logging?
> > The separate commit will essentially just rename the macros in those mpi functions that immediately get deleted in the next commit. If you insist, I'll do it.
> 
> In rlog, all are redirected to the same func. Maybe we can safely remove them. But should the change be in a sperate commit?
> 

Since there is zero impact removing the specific logging macros, I think we can skip the separation. The separation essentially is adding work that gets immediately removed, with no practical purpose at all.

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    .seealso: MPI_Accumulate
+    .impl: mpid
+    .earlyreturn: pt2pt_proc_null
+/*
```
> `request_ptr` is declared locally inside `if (unlikely(target_rank == MPI_PROC_NULL) {`, right?
My bad, I somehow saw request_ptr is accessed in fn_exit by mistake. Please ignore this comment

minsii commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,588 @@
+    The info argument provides optimization hints to the runtime about the expected
+    usage pattern of the window. The following info keys are predefined.
+    
+    . no_locks - If set to true, then the implementation may assume that passive
```
I generated the files again locally. Except a few issues that we just discussed early today, I think the PR is ready to go. One minor comment is that there are lines with only whitespace in the `Notes` area of `src/binding/c/rma_api.txt`. Please cleanup them.

hzhou replied:
Fixed.

hzhou commented on src/binding/c/rma_api.txt:
```diff
@@ -0,0 +1,466 @@
+    .skip: ThreadSafe
+    .seealso: MPI_Rget_accumulate MPI_Fetch_and_op
+    .impl: mpid
+    .earlyreturn: pt2pt_proc_null
```
Fixed.

hzhou commented:
Rebased. Removed trailing whitespace in notes. Added note for `MPI_Win_query_shared`. Patched python script to always check datatype in `rma` functions.

test:mpich/ch4/most
test:mpich/ch3/most
